stacks-signer/src/chainstate.rs

```diff
@@ -0,0 +1,375 @@
+use blockstack_lib::chainstate::nakamoto::NakamotoBlock;
+use blockstack_lib::chainstate::stacks::TenureChangePayload;
+use blockstack_lib::net::api::getsortition::SortitionInfo;
+use slog::{slog_info, slog_warn};
+use stacks_common::types::chainstate::{ConsensusHash, StacksPublicKey};
+use stacks_common::util::hash::Hash160;
+use stacks_common::{info, warn};
+
+use crate::client::{ClientError, StacksClient};
+use crate::signerdb::SignerDb;
+
+/// Captures this signer's current view of a sortition's miner.
+#[derive(PartialEq, Eq)]
+pub enum SortitionMinerStatus {
+    /// The signer thinks this sortition's miner is invalid, and hasn't signed any blocks for them.
+    InvalidatedBeforeFirstBlock,
+    /// The signer thinks this sortition's miner is invalid, but already signed one or more blocks for them.
+    InvalidatedAfterFirstBlock,
+    /// The signer thinks this sortition's miner is valid
+    Valid,
+}
+
+/// Captures the Stacks sortition related state for
+///  a successful sortition
+pub struct SortitionState {
```

Comment:
(Note to myself for review; may delete later)

It's of paramount importance that a 1-block Bitcoin reorg does not cause this struct to become structurally invalid.  Namely, we'll want to be absolutely certain that the following invariants hold:

* `prior_sortition`, `parent_tenure_id`, and `consensus_hash` correspond to sortitions which are all ancestors of the node's canonical burnchain tip
* `miner_pkh` originates from a `LeaderKeyRegisterOp` which was processed in a sortition that is an ancestor of the node's canonical burnchain tip
* The sortition containing `miner_pkh` must be an ancestor of the sortition identified by `consensus_hash`
* Either `parent_tenure_id == prior_sortition`, OR, `parent_tenure_id` is an ancestor of `prior_sortition`
* `parent_tenure_id` corresponds to a (strict) ancestor sortition of the sortition identified by `consensus_hash`